notify-debouncer-mini/src/lib.rs

```diff
@@ -160,88 +158,126 @@ pub struct DebouncedEvent {
 }
 
 impl DebouncedEvent {
+    #[inline(always)]
     fn new(path: PathBuf, kind: DebouncedEventKind) -> Self {
         Self { path, kind }
     }
 }
 
-type DebounceData = Arc<Mutex<DebounceDataInner>>;
+enum InnerEvent {
+    NotifyEvent(Result<Event, Error>),
+    Shutdown,
+}
 
-#[derive(Default)]
 struct DebounceDataInner {
+    /// Path -> Event data
     d: HashMap<PathBuf, EventData>,
+    /// timeout used to compare all events against, config
     timeout: Duration,
-    e: Vec<crate::Error>,
+    /// last debounce run
+    last_run: Instant,
+    /// oldest buffered event
+    oldest_update: Instant,
 }
 
 impl DebounceDataInner {
+    pub fn new(timeout: Duration) -> Self {
+        let time = Instant::now();
+        Self {
+            timeout: timeout,
+            oldest_update: time,
+            last_run: time,
+            d: Default::default(),
+        }
+    }
+
+    /// Returns a duration to wait for the next tick
+    #[inline]
+    pub fn next_tick(&self) -> Option<Duration> {
+        if self.has_tick_data() {
+            let next_run = self.oldest_update.checked_add(self.timeout)
+                .map_or(self.timeout, |v|v.saturating_duration_since(Instant::now()));
+            Some(next_run)
+        } else {
+            None
+        }
+    }
```

Comment:
The code looks a bit more complicate than it should. I'd suggest to:

Compute the next debounce time in `Instant` once you receive an event, so whether you should debounce or not. The computation is simply `Instant::now() + self.timeout`(that can't fail). You're already half doing so, but you store `oldest_update` but should store `debounce_deadline`.

Then the total code will look like:

```rust
pub fn next_tick(&self) -> Option<Duration> {
  self.debounce_deadline.map(|deadline| deadline.saturating_duration_since(Instant::now())
}
```

Reply:
Yeah that's a good idea.